### ISSUE TITLE
Mask Input, non ASCII input remains displayed and ignores tokens

### ISSUE THREAD

**Issue by leewes** [formkit/formkit#1325]
### Reproduction

https://formkit.com/playground?fkv=latest&fileTab=select-fill.vue&css-framework=genesis&imports=jc%28%27name%21%27ImportMap%27%7Elang%21%27json%27%7Eeditor%21%27%28*+1vue%5C%211https%3A%2F%2Fcdn.jsdelivr.net%2Fnpm%2Fvue%403%2Fdist%2Fvue.esm-browser.min.js0*%29*%27%29*%5Cn0%5C%271+0%0110*_&files=jc%5B%28%27name%21%27select-fill.vue%27%7Eeditor%21%27%3CX*%3CFormKit*6v-modelDvYRtypeDm_Rm_D%28%23%7Crepeat%29RlabelDAlphanumeric+M_+InputRhelpDTry+enteringBs+with+IME%2C+inpuQdisplay+behavior+is+erraticR%3AtokensD%28*66%22%23%22J%28*666patternJ%2F%5BA-Za-z0-9%5D%2F%2C*66%29%2C*6%29R%40keydownDURshow-m_DfalseRplaceholderDe.g.+%E5%85%A5%E5%8A%9B123457*%2F%3EHOLasQKey+pressedJ%28%28+N8OM_BJ%28%28B8OVY+lengthJ%28%28B%3F.length8HWXHH%3CscripQsetup%3E*imporQ%28+ref+%29+from+7vue7%3B*jtBSjQNSHjQUG%7Bevent%7D+%3D%3E+%28*6N.vYGevent.code*%29HWscript%3EH%27%7Esyntax%21%27html%27%7Eadded%21true%29%5D*H66++7%5C%278+%29%29WprqB+vYD%3D7G+%3D+H%5CnJ%3A+NlastKeyCodeO*%3CprqQt+R7*6SGref%7B%7D%3BUtrackKeyPressW%3C%2FXtemplatqYalue_askj*consqe%3E%01qj_YXWUSRQONJHGDB876*_

### Describe the bug

I'm currently working with the `mask` input to disallow foreign characters, but the behavior is unexpected compared to other inputs like in `number`. When inputting a non ASCII character into a `mask` input, the token does not prevent the input from displaying and has erratic IME behavior. 

**Observed behavior:**
Invalid character(s), like non ASCII / JA strings, are not removed upon input. The IME input is also partially committed to the `value`.

https://github.com/formkit/formkit/assets/90494858/639a3079-907e-4e0e-8f31-e28795558c54

Unlike `type="number"` inputs where the IME input is not partially committed and the non ASCII character(s) is removed.

https://github.com/formkit/formkit/assets/90494858/70949321-c6f0-4e5f-ad77-f061d67ebb66

**Expected behavior:**

The invalid character should be removed upon input and parts of the IME input should not be committed.

### Environment

• OS: [e.g. iOS]
Ventura 13.1 M1
• Browser [e.g. chrome, safari]
Chrome
• Version [e.g. 22]
124.0.6367.60

**Comment by justin-schroeder** [comment]
@leewes thanks for filing this! I have to admit I’m a noob when it comes to IME’s of any kind and a bit unsure how to even begin testing them. Perhaps you could assist me somewhat. The mask input uses the `beforeinput` event — do you have any thoughts on how I could simulate an IME interaction to test what kind of events we are receiving at that event?

**Comment by leewes** [comment]
> The mask input uses the beforeinput event — do you have any thoughts on how I could simulate an IME interaction to test what kind of events we are receiving at that event?

@justin-schroeder certainly! Though granted I'm also a bit new at this as well, so bear with me 😅

From what I could tell by monitoring the element, IME inputs trigger [`composition` events](https://developer.mozilla.org/en-US/docs/Web/API/Element/compositionstart_event) in addition to the `beforeInput`.

Not sure if this might help, but I created a somewhat incomplete example in sandbox:

- https://formkit.link/a0bc635421acc06a637754b7f6eb3897

that generates an composition event with the `data` of  the `v-model`-ed value of a **Dispatched Event Data** field.

**Comment by leewes** [comment]
For reference, a true ime input on a simpler string input like `aa` -> `ああ` looks like this on the events fired:

<img width="1070" alt="image" src="https://github.com/formkit/formkit/assets/90494858/cb96a659-bd9c-4760-8113-5fecd2f0a7a8">

**Comment by justin-schroeder** [comment]
Opened a Pro fix PR: https://github.com/formkit/pro/pull/83

What changed:
- Mask inputs now track the pre-composition value/selection and apply only the final composed text through the normal mask pipeline.
- Partial IME composition updates are no longer committed to the FormKit value.
- Invalid composed text, such as Japanese characters for an ASCII-only token, is removed from the visible input when composition commits.
- Valid composed text that matches the mask token still commits normally.

Verification:
- `pnpm vitest src/__tests__/maskInput.spec.ts --run -t "IME"`
- `pnpm vitest src/__tests__/maskInput.spec.ts src/__tests__/mask.spec.ts --run`
- `pnpm typecheck`
- `pnpm vitest --run`
- `pnpm build`
- `pnpm exec prettier --check src/mask.ts src/__tests__/maskInput.spec.ts`
- `git diff --check`

Security:
- No dependency, network, runtime execution, or HTML parsing changes. The change is limited to client-side mask input event handling for composition/IME events.

**Comment by justin-schroeder** [comment]
This should be fixed in `@formkit/pro@0.130.0`, released today alongside `@formkit/*` 2.1.0. Please upgrade and let us know if anything still looks off. Thanks!
